### PR TITLE
Fix Bug for Filter and Edit

### PR DESCRIPTION
**Pull request by tanyqcollin**
#191 #206 
- Fixes the issue that `Module` parameter accepts empty input and does not throw exception that cause by `Collections.emptyList()` in `Edit` and `Filter` command.

#198 #201 #226 
- Fixes the issue that `Filter` cannot filter a module that have more than one word that cause by provide wrong parameter to `StringUtill#containsWordIgnoreCase`

#208 
- Adds a description in UG that filter with empty tag will return all students.

#219 #242 
- Error mentioned in the issue 219 about the UG does not mention n/ can only be used once is not valid, since the UG has given this  instruction "Format: `filter [n/name] [p/phone] [g/gender] [t/tag]... [m/module]...`".

#204 #228 #250 #251 
- Removes the descriptions about filter phone number can accept at least 3 digit numbers.
- The constraint that recently added to `Phone` mandates phone number must have exactly 8 digits, cause the `filter` also only accepts 8-digits numbers.

#224  
- Adds specifications to the `Filter` to reduce ambiguity.
- Added `> ⚠️ **Caution:** This actions will only filter students from the data set, not from the person panel list.` to the UG.

#231 
- Adds a description to UG that indicate the `grade` is not supported to be filter.
> no find command by module as a teacher I am unable to find students that take a particular subject under me.
- This statement is not valid since this application is designed for single educator.

#235 
- Adds a warning to `Edit` command in UG that editing a module will cause grades to disappear.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/FilterCommand.java [hunk 1]
```diff
@@ -30,8 +30,9 @@ public class FilterCommand extends Command {
 
     public static final String MESSAGE_USAGE = COMMAND_WORD + ": Filter all persons whose details contain all "
             + "the specified keywords (case-insensitive) as condition and display them as  a list with index numbers.\n"
-            + "Parameters: KEYWORD [MORE_KEYWORDS]...\n"
-            + "Example: " + COMMAND_WORD + " n/joe m/MA1521";
+            + "At least one optional field must be given.\n"
+            + "Parameters: [n/NAME] [p/PHONE] [g/GENDER] [t/TAG]... [m/MODULE]...\n"
+            + "Example: " + COMMAND_WORD + " n/Alex m/Chemistry";
```
Comment (Jianwen0451) [inline src/main/java/seedu/address/logic/commands/FilterCommand.java:35]: I think this part is not a bug fix. It is a feature tweak as it is trying to make the error message of FilterCommand more specific. Please Undo this part

**Review by Jianwen0451 — CHANGES_REQUESTED** [review]
The issue #242 is a feature flaw (Error message not clear), rather than a functionality bug. Please undo the change of the error message

**Comment by Jianwen0451** [comment]
Also Fixing #208 also seems not appropriate. 

Screenshot 2024-11-09 at 3 35 38 PM

As stated here, the garbage-in garbage-out is not necessarily "incorrect" and can be considered as a feature update.

My suggestion is to undo the change and just say in UG that enter empty tags will match for all students

**Comment by tanyqcollin** [comment]
> Also Fixing #208 also seems not appropriate.
> 
> Screenshot 2024-11-09 at 3 35 38 PM
> As stated here, the garbage-in garbage-out is not necessarily "incorrect" and can be considered as a feature update.
> 
> My suggestion is to undo the change and just say in UG that enter empty tags will match for all students

image
Previously, I noticed this issue with filtering empty tags, which I fixed on my local host in this commit https://github.com/AY2425S1-CS2103T-W10-4/tp/pull/154/commits/8959fddaf05800edc31e32e14145666b5703c561. However, after merging the branch, I did not test it again by filter with an empty tags, so I believe this might still considered as a bug.

File: src/main/java/seedu/address/logic/commands/FilterCommand.java (hunk 1 above)
Comment (tanyqcollin) [reply src/main/java/seedu/address/logic/commands/FilterCommand.java:35]: noted.

**Comment by Jianwen0451** [comment]
> > Also Fixing #208 also seems not appropriate.
> > Screenshot 2024-11-09 at 3 35 38 PM
> > As stated here, the garbage-in garbage-out is not necessarily "incorrect" and can be considered as a feature update.
> > My suggestion is to undo the change and just say in UG that enter empty tags will match for all students
> 
> image Previously, I noticed this issue with filtering empty tags, which I fixed on my local host in this commit [8959fdd](https://github.com/AY2425S1-CS2103T-W10-4/tp/commit/8959fddaf05800edc31e32e14145666b5703c561). However, after merging the branch, I did not test it again by filter with an empty tags, so I believe this might still considered as a bug.

In commit https://github.com/AY2425S1-CS2103T-W10-4/tp/commit/8959fddaf05800edc31e32e14145666b5703c561, I see you implement an orElse(true) at the end. This means that the predicate for the filter tag will be true if the give tag list is empty. This means that all persons will be listed when nothing but an empty tag list is given as input. So it is even not a wrong "output". This just means that it ***will be better*** if we disallow this and produce an error message. I am strongly against this change.  I suggest that we just play safe here, not changing it but can mention it in the DG "Planned Enhancement". The cost of making a feature change is too high for all of us.

**Review by Jianwen0451 — APPROVED** [review]

**Review by yongqqqq — APPROVED** [review]
LGTM!

**Pull request merged**